dockstore-webservice/src/main/java/io/dockstore/webservice/languages/CWLHandler.java

```diff
@@ -896,4 +908,172 @@ public VersionTypeValidation validateToolSet(Set<SourceFile> sourcefiles, String
     public VersionTypeValidation validateTestParameterSet(Set<SourceFile> sourceFiles) {
         return checkValidJsonAndYamlFiles(sourceFiles, DescriptorLanguage.FileType.CWL_TEST_JSON);
     }
+
+    public static class Preprocessor {
+        private static final List<String> IMPORT_KEYS = Arrays.asList("$import", "import");
+        private static final List<String> INCLUDE_KEYS = Arrays.asList("$include", "include");
+        private static final List<String> MIXIN_KEYS = Arrays.asList("$mixin", "mixin");
+        private static final int DEFAULT_MAX_DEPTH = 8;
+
+        private final Set<SourceFile> secondarySourceFiles;
+        private final int maxDepth;
+        private final Map<String, String> idToPath;
+
+        public Preprocessor(Set<SourceFile> secondarySourceFiles, int maxDepth) {
+            this.secondarySourceFiles = secondarySourceFiles;
+            this.maxDepth = maxDepth;
+            this.idToPath = new HashMap<>();
+        }
+
+        public Preprocessor(Set<SourceFile> secondarySourceFiles) {
+            this(secondarySourceFiles, DEFAULT_MAX_DEPTH);
+        }
+
+        private void preprocessMapValues(Map<String, Object> map, String currentPath, int depth) {
+            map.replaceAll((k, v) -> preprocess(v, currentPath, depth));
+            Object runValue = map.get("run");
+            if (runValue instanceof String) {
+                LOG.error("RUN " + currentPath + " " + depth + ": " + runValue);
+                String runPath = (String)runValue;
+                map.put("run", loadFileAndPreprocess(resolvePath(runPath, currentPath), depth));
+            }
+        }
+
+        private void preprocessListValues(List<Object> list, String currentPath, int depth) {
+            list.replaceAll(v -> preprocess(v, currentPath, depth));
+        }
+
+        private boolean isTool(Map<String, Object> map) {
+            return Objects.equals(map.get("class"), "CommandLineTool");
+        }
+
+        private String addId(Map<String, Object> map) {
+            map.putIfAbsent("id", java.util.UUID.randomUUID().toString());
+            return (String)map.get("id"); // TODO fix?
+        }
+
+        public Object preprocess(Object obj, String currentPath, int depth) {
```

Comment:
think at least the public methods need commenting, think this is manipulating a map of workflow steps?